### PR TITLE
Jsonb support for postgres

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

**Comment by YourTechBud** [comment]
Ypur tests are failing

File: gateway/model/schema_type.go [hunk 1]
```diff
@@ -56,6 +56,8 @@ const (
 	TypeDateTime string = "DateTime"
 	//TypeID is variable used for Variable of type ID
 	TypeID string = "ID"
+	// TypeJsonb is variable used for Variable of type Jsonb
+	TypeJsonb string = "Jsonb"
```
Comment (YourTechBud) [inline gateway/model/schema_type.go]:
```suggestion
	TypeJsonb string = "JSON"
```

File: gateway/modules/auth/post_action.go [hunk 2]
```diff
@@ -23,14 +23,17 @@ func (m *Module) PostProcessMethod(postProcess *model.PostProcess, result interf
...
 	case []interface{}:
 		resultArr = val
 	default:
 		return errors.New("result is of invalid type")
 	}
 
 	for _, doc := range resultArr {
+		if postProcess == nil {
+			break
+		}
```
Comment (YourTechBud) [inline gateway/modules/auth/post_action.go]: Do this when we check `result`. Put an `||` clause there

File: gateway/modules/crud/crud.go [hunk 3]
```diff
@@ -28,6 +28,11 @@ type Module struct {
 	project            string
 	removeProjectScope bool
 
+	// during a read request the result contains a column having type json for (postgres)
+	// then un marshall the json field in read request
+	// Note : reason for having field type func as we cannot directly import schema module in crud
```
Comment (YourTechBud) [inline gateway/modules/crud/crud.go]: Start with capital letter. The grammar seems to bit off

File: gateway/modules/crud/operations.go [hunk 4]
```diff
@@ -52,6 +53,11 @@ func (m *Module) Read(ctx context.Context, dbAlias, project, col string, req *mo
 	m.RLock()
 	defer m.RUnlock()
 
+	dbType, err := m.GetDBType(dbAlias)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (YourTechBud) [inline gateway/modules/crud/operations.go]: No need of doing this. We you can use `crud.Type()` instead.

File: gateway/modules/crud/sql/update.go [hunk 5]
```diff
@@ -4,13 +4,14 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/sirupsen/logrus"
```
Comment (YourTechBud) [inline gateway/modules/crud/sql/update.go]: You cannot import this along with internal packages. Put it down.

File: gateway/modules/schema/crud.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		docs = []interface{}{v}
+	}
+	for columnName, columnValue := range tableInfo {
+		if columnValue.Kind == model.TypeJsonb {
+			for _, doc := range docs {
+				finalDoc := doc.(map[string]interface{})
+				data, ok := finalDoc[columnName].([]byte)
+				if !ok {
+					logrus.Errorf("error crud post process in schema module unable to type assert interface to []byte for column (%s)", columnName)
+					return fmt.Errorf("unable to type assert interface to []byte")
```
Comment (YourTechBud) [inline gateway/modules/schema/crud.go]: Include the field name too here

File: gateway/modules/schema/crud.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+				}
+				var v interface{}
+				if err := json.Unmarshal(data, &v); err != nil {
+					logrus.Errorf("error crud post process in schema module unable unmarshal data (%s)", string(data))
+					return fmt.Errorf("unable to unmarshal json data")
```
Comment (YourTechBud) [inline gateway/modules/schema/crud.go]: Include the field name too here

File: gateway/modules/schema/helpers.go [hunk 8]
```diff
@@ -35,11 +35,17 @@ func getSQLType(dbType, typename string) (string, error) {
...
+		if dbType != string(utils.Postgres) {
+			return "", fmt.Errorf("jsonb not supported for database %s", dbType)
+		}
+		return "jsonb", nil
 	default:
 		return "", fmt.Errorf("%s type not allowed", typename)
 	}
 }
 
+// TODO ANY PARTICULAR DIRECTIVE THAT IS NOT SUPPORTED FOR JSONB
```
Comment (YourTechBud) [inline gateway/modules/schema/helpers.go]: What is this for?

File: gateway/modules/schema/inspection.go [hunk 9]
```diff
@@ -172,7 +172,8 @@ func inspectionPostgresCheckFieldType(typeName string, fieldDetails *model.Field
 		fieldDetails.Kind = model.TypeDateTime
 	case "boolean":
 		fieldDetails.Kind = model.TypeBoolean
-
+	case "jsonb":
```
Comment (YourTechBud) [inline gateway/modules/schema/inspection.go]:
```suggestion
	case "jsonb", "json":
```

File: gateway/modules/schema/validatecreate.go [hunk 10]
```diff
@@ -1,10 +1,13 @@
 package schema
 
 import (
+	"encoding/json"
 	"errors"
 	"fmt"
 	"time"
 
+	"github.com/sirupsen/logrus"
+
```
Comment (YourTechBud) [inline gateway/modules/schema/validatecreate.go]: Empty space not rquired

File: gateway/utils/server/server.go [hunk 11]
```diff
@@ -195,6 +195,8 @@ func (s *Server) LoadConfig(config *config.Config) error {
 			return err
 		}
 
+		s.crud.SetSchemaPostProcessHandler(s.schema.CrudPostProcess)
```
Comment (YourTechBud) [inline gateway/utils/server/server.go]: Do this in the `New()` function.

File: gateway/utils/validate.go [hunk 12]
```diff
@@ -185,7 +185,8 @@ func Validate(where map[string]interface{}, obj interface{}) bool {
 					default:
 						return false
 					}
-
+				case "$contains":
+					// TODO : I DON'T KNOW IF IT IS NECCESSARY
```
Comment (YourTechBud) [inline gateway/utils/validate.go]: Yes you need to do this.

File: gateway/modules/schema/helpers.go (hunk 8 above)
Comment (sharadregoti) [reply gateway/modules/schema/helpers.go]: i have not tested jsob with any particular directive like adding @primary or @unique for a field type having json
so the question is does JSON supports all directives ?
if no then what directives are not supported
we can throw err while checking for error in schema creation

Comment (YourTechBud) [reply gateway/modules/schema/helpers.go]: Create an issue for this. Indexes on jsonb types

Comment (sharadregoti) [reply gateway/modules/schema/helpers.go]: ok

Comment (sharadregoti) [reply gateway/modules/schema/helpers.go]: so should i throw error now if the user tries to create index on jsonb types

Comment (YourTechBud) [reply gateway/modules/schema/helpers.go]: Yes

File: gateway/utils/validate.go [hunk 13]
```diff
@@ -185,7 +185,7 @@ func Validate(where map[string]interface{}, obj interface{}) bool {
 					default:
 						return false
 					}
-
+				case "$contains":
```
Comment (YourTechBud) [inline gateway/utils/validate.go]: You need to write the validate logic here.

File: gateway/model/schema_type.go [hunk 14]
```diff
@@ -56,7 +56,9 @@ const (
 	TypeDateTime string = "DateTime"
 	//TypeID is variable used for Variable of type ID
 	TypeID string = "ID"
-	//SQLTypeIDSize is variable used for specifing size of sql type ID
+	// TypeJsonb is variable used for Variable of type Jsonb
+	TypeJsonb string = "JSON"
```
Comment (YourTechBud) [inline gateway/model/schema_type.go]: TypeJSON

**Comment by sharadregoti** [comment]
yes

File: gateway/utils/validate.go (hunk 13 above)
Comment (YourTechBud) [reply gateway/utils/validate.go]: The logic here is incorrect. You need to actually implement the contains quiet here. Think of this function as a filter

File: gateway/utils/validate.go [hunk 15]
```diff
@@ -186,9 +186,28 @@ func Validate(where map[string]interface{}, obj interface{}) bool {
...
+						result, ok := res[k]
+						if !ok {
+							return false
+						}
+						resultObj, ok := result.(map[string]interface{})
+						if !ok {
+							return false
+						}
+						// check if the top level key values are same
+						for key, value := range v {
```
Comment (YourTechBud) [inline gateway/utils/validate.go]: This logic is incorrect. You need to check if all top level fields are equal. Correctly you are returning true even if one matches

File: gateway/utils/validate_test.go [hunk 16]
```diff
@@ -356,6 +356,38 @@ func TestValidate(t *testing.T) {
...
+		{
+			name: "valid contains",
+			args: args{
+				where: map[string]interface{}{"op2": map[string]interface{}{"$contains": map[string]interface{}{"a": 1, "b": "one", "c": true,}}},
+				obj:   map[string]interface{}{"op2": map[string]interface{}{"a": 1, "b": "one", "c": true,}},
+			},
+			want: true,
+		},
+		{
+			name: "invalid contains",
```
Comment (YourTechBud) [inline gateway/utils/validate_test.go]: Add one more test case, what if some keys match and some don't

File: gateway/modules/crud/crud.go [hunk 17]
```diff
@@ -28,6 +28,12 @@ type Module struct {
 	project            string
 	removeProjectScope bool
 	schema             model.SchemaCrudInterface
+
+	// During a read request if the result contains a column of type jsonb whose value is stored as []bytes
+	// Then un marshall that particular field of the result & override the value with un marshalling result
+	// Note : reason for having this field here is that we cannot directly import schema module in crud
+	schemaPostProcess func(ctx context.Context, dbAlias, col string, result interface{}) error
```
Comment (YourTechBud) [inline gateway/modules/crud/crud.go]: No need to do this since you have schema in crud

File: gateway/utils/validate.go (hunk 15 above)
Comment (sharadregoti) [reply gateway/utils/validate.go]: that's how it works in postgres. that' why i wrote the logic like that.

File: gateway/modules/schema/update.go [hunk 18]
```diff
@@ -15,36 +17,44 @@ func (s *Schema) ValidateUpdateOperation(dbAlias, col, op string, updateDoc, fin
...
 	}
 	SchemaDoc, ok := schemaDb[col]
 	if !ok {
+		logrus.Infof("validating update operation in schema module collection (%s) not found in schemaDoc where dbAlias (%s)", col, dbAlias)
 		return nil
 	}
 
 	for key, doc := range updateDoc {
 		switch key {
+		case "$unset":
```
Comment (YourTechBud) [inline gateway/modules/schema/update.go]: Unset should be allowed only on non required fields for mongo, and JSONB fields for postgres. MySQL and Sqlserver not allowed

File: gateway/utils/validate_test.go (hunk 16 above)
Comment (sharadregoti) [reply gateway/utils/validate_test.go]: not required

File: gateway/utils/server/server.go [hunk 19]
```diff
@@ -70,6 +70,8 @@ func New(nodeID, clusterID, advertiseAddr, storeType, runnerAddr, artifactAddr s
 	a := auth.Init(nodeID, c, removeProjectScope)
 	a.SetMakeHTTPRequest(syncMan.MakeHTTPRequest)
 
+	c.SetSchemaPostProcessHandler(s.CrudPostProcess)
```
Comment (YourTechBud) [inline gateway/utils/server/server.go]: Won't be required

File: gateway/utils/validate.go (hunk 15 above)
Comment (YourTechBud) [reply gateway/utils/validate.go]: Postgres matches a doc even if one key matches? And what about nested fields?

File: gateway/utils/validate_test.go (hunk 16 above)
Comment (YourTechBud) [reply gateway/utils/validate_test.go]: Add a test case. Whats wrong in that

File: gateway/modules/schema/inspection_test.go [hunk 20]
```diff
@@ -95,18 +95,29 @@ func Test_generateInspection(t *testing.T) {
...
 				col:         "table1",
-				fields:      []utils.FieldType{utils.FieldType{FieldName: "col1", FieldType: "wrongType", FieldNull: "NO", FieldKey: "MUL"}},
+				fields:      []utils.FieldType{{FieldName: "col1", FieldType: "wrongType", FieldNull: "NO", FieldKey: "MUL"}},
 				foreignkeys: []utils.ForeignKeysType{utils.ForeignKeysType{TableName: "table1", ColumnName: "col1", RefTableName: "table2", RefColumnName: "col2"}},
 			},
 			wantErr: true,
 		},
 		// postgres
+		{
+			name: "JSON with not null",
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply gateway/modules/schema/inspection_test.go]: yes. It doen't generate any queries when there are no changes

File: gateway/utils/validate.go (hunk 15 above)
Comment (sharadregoti) [reply gateway/utils/validate.go]: yes, same applies for nested fields

Comment (sharadregoti) [reply gateway/utils/validate.go]: image

Comment (sharadregoti) [reply gateway/utils/validate.go]: do i use recursion for checking the nested fields ?
or is just enough to compare to interface using "=="

Comment (YourTechBud) [reply gateway/utils/validate.go]: Use recursion

File: gateway/modules/schema/update.go [hunk 21]
```diff
@@ -164,6 +176,43 @@ func validateDateOperations(col string, doc interface{}, SchemaDoc model.Fields)
...
+	}
+	if dbType == string(utils.Mongo) {
+		for fieldName := range v {
+			columnInfo, ok := schemaDoc[fieldName]
+			if ok {
+				if columnInfo.IsFieldTypeRequired {
+					return fmt.Errorf("cannot use $unset on field which is required")
+				}
+			}
+			// TODO CAN SPECIFIED FIELD NAME DOESN'T EXIST IN SCHEMA IN CASE OF MONGO
```
Comment (YourTechBud) [inline gateway/modules/schema/update.go]: Yes this is allowed

File: gateway/modules/schema/update.go [hunk 22]
```diff
@@ -164,6 +176,43 @@ func validateDateOperations(col string, doc interface{}, SchemaDoc model.Fields)
...
+	if !ok {
+		return fmt.Errorf("document not of type object in collection %s", col)
+	}
+	dbType, err := s.crud.GetDBType(dbAlias)
+	if err != nil {
+		return err
+	}
+	if dbType == string(utils.Mongo) {
+		for fieldName := range v {
+			columnInfo, ok := schemaDoc[fieldName]
```
Comment (YourTechBud) [inline gateway/modules/schema/update.go]: Fieldname will be in the form of `f1.f2.f3`

File: gateway/modules/schema/update.go [hunk 23]
```diff
@@ -164,6 +176,43 @@ func validateDateOperations(col string, doc interface{}, SchemaDoc model.Fields)
...
+				}
+			}
+			// TODO CAN SPECIFIED FIELD NAME DOESN'T EXIST IN SCHEMA IN CASE OF MONGO
+		}
+		return nil
+	}
+
+	if dbType == string(utils.Postgres) || dbType == string(utils.MySQL) || dbType == string(utils.SQLServer) {
+		for fieldName := range v {
+			columnInfo, ok := schemaDoc[fieldName]
```
Comment (YourTechBud) [inline gateway/modules/schema/update.go]: Fieldname is in the form of `f1.f2.f3`. only the top level field will be in the schema

File: gateway/utils/validate_test.go [hunk 24]
```diff
@@ -388,6 +388,104 @@ func TestValidate(t *testing.T) {
...
+			args: args{
+				where: map[string]interface{}{"op2": map[string]interface{}{"$contains": map[string]interface{}{"foo2": map[string]interface{}{"inner5": []interface{}{}}}}},
+				obj:   map[string]interface{}{"op2": map[string]interface{}{"foo1": "bar1", "foo2": map[string]interface{}{"inner1": "value1", "inner2": true, "inner3": 1.4, "inner4": 4, "inner5": []interface{}{1, 2, 3}}}},
+			},
+			want: true,
+		},
+		{
+			name: "valid contains nested array contains value",
+			args: args{
+				where: map[string]interface{}{"op2": map[string]interface{}{"$contains": map[string]interface{}{"foo2": map[string]interface{}{"inner5": []interface{}{1, 3}}}}},
```
Comment (YourTechBud) [inline gateway/utils/validate_test.go]: I don't think the where clause will contain an array

**Pull request merged**